File: app/skills/validate_glossary_terms.py
"""Validate glossary_terms rows for form, semantic accuracy, and dedupe.

Phase C — glossary stream reinforcement. Spec: ``design/skills/validate-glossary-terms.md``.

Three checks:

* **form** — purely Python, deterministic. Hits empty defs, over-long
  defs (>25 words / >2 sentences), banned-word infiltration into the
  quick_def line (jargon the writer was supposed to avoid).
* **semantic** — LLM-as-judge. The orchestrator hands each (term,
  quick_def, why_it_matters) tuple to the writer's LLM context
  with the prompt template in the skill spec; the LLM returns
  ``{"verdict": "match|mismatch|uncertain", "reason": ..., "suggested_fix": ...}``.
  ``commit_validation()`` writes that to ``glossary_audit`` and, on
  ``mismatch``, retires the row with ``reviewed_by_human=1``
  (so the auto-retire loop doesn't keep flapping it).
* **dedupe** — purely Python. Detects when a fresh active term is a
  case-insensitive substring or alias-overlap of another active term
  and flags the shorter one for retirement. Conservative — only fires
  when the overlap is unambiguous.
"""

from __future__ import annotations

import argparse
import datetime as dt
import json
import re
import sqlite3
import sys
from pathlib import Path


# Words that should not appear inside a quick_def sentence — these are
# the jargon the line is supposed to translate AWAY from. The list is
# intentionally narrow: too aggressive a list and every legitimate AI
# definition would trip.
_FORBIDDEN_IN_QUICK_DEF = frozenset(
    {
        "leverage", "synergy", "paradigm", "ecosystem",  # corporate noise
        "utilize", "facilitate",                         # use 'use' / 'help'
        "stochastic", "heuristic", "asymptotic",         # not 14-yr friendly
        # add as needed; intentionally short
    }
)

# Words that signal the term itself is generic English, not a proper
# noun / acronym worthy of a glossary entry. Used by the form check
# when the row's `term` column itself looks generic.
_GENERIC_TERM_WORDS = frozenset(
    {
        "today", "yesterday", "tomorrow", "future", "past",
        "news", "summary", "headlines", "report",
        "team", "company", "user", "users",
        "thing", "stuff", "issue", "case",
    }
)


def _word_count(s: str) -> int:
    return len(re.findall(r"\S+", s or ""))


def _sentence_count(s: str) -> int:
    if not s:
        return 0
    return len([p for p in re.split(r"(?<=[.!?])\s+", s.strip()) if p])


# ---------------------------------------------------------------------------
# Form check (Python only)
# ---------------------------------------------------------------------------


def form_check(row: dict) -> dict:
    """Run the form-check on one row. Returns one verdict dict.

    Verdict shape: {"check_type": "form", "verdict": "...", "reason": "...",
                    "suggested_fix": "..." (optional)}
    """
    term = (row.get("term") or "").strip()
    qd = (row.get("quick_def") or "").strip()
    why = (row.get("why_it_matters") or "").strip()
    issues: list[str] = []

    if not term:
        return {"check_type": "form", "verdict": "fail",
                "reason": "term is empty"}

    if term.lower() in _GENERIC_TERM_WORDS:
        issues.append(f"term {term!r} looks like a generic English word, not a proper noun / acronym")

    if not qd:
        return {"check_type": "form", "verdict": "fail",
                "reason": "quick_def is empty"}

    qd_words = _word_count(qd)
    if qd_words > 25:
        issues.append(f"quick_def is {qd_words} words (cap 25)")
    qd_sentences = _sentence_count(qd)
    if qd_sentences > 2:
        issues.append(f"quick_def has {qd_sentences} sentences (cap 2)")

    qd_lower = qd.lower()
    forbidden_hits = [w for w in _FORBIDDEN_IN_QUICK_DEF if re.search(rf"\b{re.escape(w)}\b", qd_lower)]
    if forbidden_hits:
        issues.append(f"quick_def contains forbidden jargon: {forbidden_hits}")

    if why:
        why_words = _word_count(why)
        if why_words > 30:
            issues.append(f"why_it_matters is {why_words} words (cap 30)")

    if not issues:
        return {"check_type": "form", "verdict": "pass", "reason": ""}
    # `fail` if the term is generic OR the quick_def has hard-blockers;
    # otherwise `warn`.
    blocking = any("generic English" in i or "forbidden jargon" in i for i in issues)
    return {
        "check_type": "form",
        "verdict": "fail" if blocking else "warn",
        "reason": "; ".join(issues),
    }


# ---------------------------------------------------------------------------
# Dedupe check (Python only — conservative)
# ---------------------------------------------------------------------------


def dedupe_check(conn: sqlite3.Connection, row: dict) -> dict:
    """Flag a row whose term + alias set is fully covered by an existing
    older active term. Conservative: only fires on exact case-insensitive
    overlap (term equals an alias of another row, or aliases overlap by
    ≥ 1 token AND the term is a substring of the other term).
    """
    term = (row.get("term") or "").strip()
    if not term:
        return {"check_type": "dedupe", "verdict": "pass", "reason": ""}
    aliases = []
    raw_aliases = row.get("aliases_json") or "[]"
    try:
        a = json.loads(raw_aliases)
        if isinstance(a, list):
            aliases = [str(x).strip() for x in a if x]
    except (json.JSONDecodeError, TypeError):
        aliases = []
    own = {term.lower(), *(a.lower() for a in aliases)}
    cur = conn.execute(
        """
        SELECT term, aliases_json FROM glossary_terms
         WHERE status = 'active' AND term <> ?
        """,
        (term,),
    )
    for other in cur.fetchall():
        other_term = other["term"]
        other_aliases = []
        try:
            o = json.loads(other["aliases_json"] or "[]")
            if isinstance(o, list):
                other_aliases = [str(x).strip() for x in o if x]
        except (json.JSONDecodeError, TypeError):
            pass
        other_set = {other_term.lower(), *(a.lower() for a in other_aliases)}
        # If our term is in their alias set OR fully overlaps, flag.
        if term.lower() in other_set:
            return {
                "check_type": "dedupe",
                "verdict": "fail",
                "reason": f"term {term!r} is already an alias of active term {other_term!r}",
                "suggested_fix": f"retire and add to {other_term!r} aliases_json",
            }
        if own & other_set and len(other_term) > len(term) and term.lower() in other_term.lower():
            return {
                "check_type": "dedupe",
                "verdict": "warn",
                "reason": f"term {term!r} overlaps with longer active term {other_term!r}",
                "suggested_fix": f"consider merging into {other_term!r}",
            }
    return {"check_type": "dedupe", "verdict": "pass", "reason": ""}


# ---------------------------------------------------------------------------
# Pending semantic checks — surface terms the LLM-as-judge needs to look at
# ---------------------------------------------------------------------------


def list_pending_semantic(conn: sqlite3.Connection, *, limit: int = 25) -> list[dict]:
    """Return active rows that haven't had a passing semantic audit yet."""
    cur = conn.execute(
        """
        SELECT t.term, t.aliases_json, t.quick_def, t.why_it_matters,
               t.canonical_link
          FROM glossary_terms t
         WHERE t.status = 'active'
           AND t.quick_def IS NOT NULL
           AND t.reviewed_by_human = 0
           AND NOT EXISTS (
             SELECT 1 FROM glossary_audit a
              WHERE a.term = t.term
                AND a.check_type = 'semantic'
                AND a.verdict IN ('pass', 'fail')
           )
         ORDER BY t.term
         LIMIT ?
        """,
        (limit,),
    )
    return [dict(r) for r in cur.fetchall()]


# ---------------------------------------------------------------------------
# Audit log + auto-retire
# ---------------------------------------------------------------------------


def commit_validation(
    conn: sqlite3.Connection,
    *,
    term: str,
    verdicts: list[dict],
) -> None:
    """Write each verdict to glossary_audit. If any verdict is 'fail',
    flip the row to status='retired' and reviewed_by_human=1 (so the
    auto-retire-quiet loop in define-glossary-terms doesn't churn on
    it again next run)."""
    today = dt.date.today().isoformat()
    has_fail = False
    for v in verdicts:
        conn.execute(
            """
            INSERT INTO glossary_audit (term, check_type, verdict, reason, suggested_fix, checked_at)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (
                term,
                v.get("check_type", "form"),
                v.get("verdict", "pass"),
                v.get("reason"),
                v.get("suggested_fix"),
                today,
            ),
        )
        if v.get("verdict") == "fail":
            has_fail = True
    if has_fail:
        conn.execute(
            """
            UPDATE glossary_terms
               SET status = 'retired',
                   reviewed_by_human = 1,
                   updated_at = ?
             WHERE term = ?
            """,
            (today, term),
        )
    conn.commit()


def run(db_path: Path, *, limit: int) -> dict:
    if not db_path.is_file():
        raise FileNotFoundError(f"DB not found: {db_path}")
    conn = sqlite3.connect(db_path)
    conn.row_factory = sqlite3.Row
    try:
        cur = conn.execute(
            """
            SELECT term, aliases_json, quick_def, why_it_matters, canonical_link
              FROM glossary_terms
             WHERE status = 'active'
             ORDER BY term
             LIMIT ?
            """,
            (limit,),
        )
        rows = [dict(r) for r in cur.fetchall()]
        results: list[dict] = []
        retired: list[str] = []
        warns: list[str] = []
        for row in rows:
            verdicts = [
                form_check(row),
                dedupe_check(conn, row),
            ]
            commit_validation(conn, term=row["term"], verdicts=verdicts)
            for v in verdicts:
                if v["verdict"] == "fail":
                    retired.append(row["term"])
                    break
            else:
                if any(v["verdict"] == "warn" for v in verdicts):
                    warns.append(row["term"])
            results.append({"term": row["term"], "verdicts": verdicts})
        pending_semantic = [r["term"] for r in list_pending_semantic(conn, limit=limit)]
    finally:
        conn.close()
    return {
        "checked": len(rows),
        "retired_by_form_or_dedupe": retired,
        "warned": warns,
        "pending_semantic_count": len(pending_semantic),
        "pending_semantic": pending_semantic,
        "results": results,
    }


def main(argv: list[str] | None = None) -> int:
    p = argparse.ArgumentParser(description="Validate active glossary terms (form + dedupe + queue semantic)")
    p.add_argument("--db", required=True, type=Path)
    p.add_argument("--limit", type=int, default=200)
    args = p.parse_args(argv)
    try:
        out = run(args.db, limit=args.limit)
    except FileNotFoundError as e:
        print(f"FAIL {e}", file=sys.stderr)
        return 2
    print(json.dumps(out, indent=2, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: engines/nunc-fluens/pipeline/src/gates/validate-glossary-terms.ts
// TS port of app/skills/validate_glossary_terms.py — form (det),
// dedupe (det), and the semantic LLM-as-judge queue + audit-log write.
// Every verdict lands in glossary_audit; a 'fail' retires the row with
// reviewed_by_human=1 so define-glossary-terms' quiet-rule doesn't
// churn on it. Semantic verdicts map match→pass, mismatch→fail,
// uncertain→warn (the audit CHECK constraint owns the vocabulary; a
// 'warn' row keeps the term in the pending-semantic queue).
import type { Db } from '../ingest/ingest-core.ts';

const FORBIDDEN_IN_QUICK_DEF = new Set([
  'leverage', 'synergy', 'paradigm', 'ecosystem',
  'utilize', 'facilitate',
  'stochastic', 'heuristic', 'asymptotic',
]);

const GENERIC_TERM_WORDS = new Set([
  'today', 'yesterday', 'tomorrow', 'future', 'past',
  'news', 'summary', 'headlines', 'report',
  'team', 'company', 'user', 'users',
  'thing', 'stuff', 'issue', 'case',
]);

export interface GlossaryRow {
  term: string;
  aliases_json: string | null;
  quick_def: string | null;
  why_it_matters: string | null;
  canonical_link: string | null;
}

export interface Verdict {
  check_type: 'form' | 'semantic' | 'dedupe';
  verdict: 'pass' | 'warn' | 'fail';
  reason: string;
  suggested_fix?: string;
}

function wordCount(s: string): number {
  return (s.match(/\S+/g) ?? []).length;
}

function sentenceCount(s: string): number {
  if (!s) return 0;
  return s.trim().split(/(?<=[.!?])\s+/).filter(p => p).length;
}

function escapeRe(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export function formCheck(row: GlossaryRow): Verdict {
  const term = (row.term ?? '').trim();
  const qd = (row.quick_def ?? '').trim();
  const why = (row.why_it_matters ?? '').trim();
  const issues: string[] = [];

  if (!term) return { check_type: 'form', verdict: 'fail', reason: 'term is empty' };
  if (GENERIC_TERM_WORDS.has(term.toLowerCase()))
    issues.push(`term '${term}' looks like a generic English word, not a proper noun / acronym`);
  if (!qd) return { check_type: 'form', verdict: 'fail', reason: 'quick_def is empty' };

  const qdWords = wordCount(qd);
  if (qdWords > 25) issues.push(`quick_def is ${qdWords} words (cap 25)`);
  const qdSentences = sentenceCount(qd);
  if (qdSentences > 2) issues.push(`quick_def has ${qdSentences} sentences (cap 2)`);

  const qdLower = qd.toLowerCase();
  const forbiddenHits = [...FORBIDDEN_IN_QUICK_DEF]
    .filter(w => new RegExp(`\\b${escapeRe(w)}\\b`).test(qdLower));
  if (forbiddenHits.length)
    issues.push(`quick_def contains forbidden jargon: ${forbiddenHits.join(', ')}`);

  if (why) {
    const whyWords = wordCount(why);
    if (whyWords > 30) issues.push(`why_it_matters is ${whyWords} words (cap 30)`);
  }

  if (!issues.length) return { check_type: 'form', verdict: 'pass', reason: '' };
  const blocking = issues.some(i =>
    i.includes('generic English') || i.includes('forbidden jargon'));
  return { check_type: 'form', verdict: blocking ? 'fail' : 'warn', reason: issues.join('; ') };
}

function parseAliases(raw: string | null): string[] {
  try {
    const a = JSON.parse(raw ?? '[]');
    return Array.isArray(a) ? a.filter(x => x).map(x => String(x).trim()) : [];
  } catch {
    return [];
  }
}

export function dedupeCheck(db: Db, row: GlossaryRow): Verdict {
  const term = (row.term ?? '').trim();
  if (!term) return { check_type: 'dedupe', verdict: 'pass', reason: '' };
  const own = new Set([term.toLowerCase(),
    ...parseAliases(row.aliases_json).map(a => a.toLowerCase())]);
  const others = db.prepare(
    `SELECT term, aliases_json FROM glossary_terms
      WHERE status = 'active' AND term <> ?`,
  ).all(term) as Array<{ term: string; aliases_json: string | null }>;
  for (const other of others) {
    const otherSet = new Set([other.term.toLowerCase(),
      ...parseAliases(other.aliases_json).map(a => a.toLowerCase())]);
    if (otherSet.has(term.toLowerCase())) {
      return {
        check_type: 'dedupe', verdict: 'fail',
        reason: `term '${term}' is already an alias of active term '${other.term}'`,
        suggested_fix: `retire and add to '${other.term}' aliases_json`,
      };
    }
    const overlaps = [...own].some(x => otherSet.has(x));
    if (overlaps && other.term.length > term.length
      && other.term.toLowerCase().includes(term.toLowerCase())) {
      return {
        check_type: 'dedupe', verdict: 'warn',
        reason: `term '${term}' overlaps with longer active term '${other.term}'`,
        suggested_fix: `consider merging into '${other.term}'`,
      };
    }
  }
  return { check_type: 'dedupe', verdict: 'pass', reason: '' };
}

export function listPendingSemantic(db: Db, limit = 25): GlossaryRow[] {
  return db.prepare(
    `SELECT t.term, t.aliases_json, t.quick_def, t.why_it_matters,
            t.canonical_link
       FROM glossary_terms t
      WHERE t.status = 'active'
        AND t.quick_def IS NOT NULL
        AND t.reviewed_by_human = 0
        AND NOT EXISTS (
          SELECT 1 FROM glossary_audit a
           WHERE a.term = t.term
             AND a.check_type = 'semantic'
             AND a.verdict IN ('pass', 'fail')
        )
      ORDER BY t.term
      LIMIT ?`,
  ).all(limit) as GlossaryRow[];
}

export function commitValidation(db: Db, args: {
  term: string; verdicts: Verdict[]; today: string;
}): void {
  const ins = db.prepare(
    `INSERT INTO glossary_audit (term, check_type, verdict, reason, suggested_fix, checked_at)
     VALUES (?, ?, ?, ?, ?, ?)`);
  let hasFail = false;
  for (const v of args.verdicts) {
    ins.run(args.term, v.check_type, v.verdict, v.reason ?? null,
      v.suggested_fix ?? null, args.today);
    if (v.verdict === 'fail') hasFail = true;
  }
  if (hasFail) {
    db.prepare(
      `UPDATE glossary_terms
          SET status = 'retired', reviewed_by_human = 1, updated_at = ?
        WHERE term = ?`,
    ).run(args.today, args.term);
  }
}

/** 30-day retention for glossary_audit (post-C decision P7).
 *
 * Semantic pass/fail rows are deliberately EXEMPT: listPendingSemantic's
 * anti-join exists to avoid re-judging terms with the LLM — pruning those
 * rows would re-queue every term for the judge each month. Form/dedupe
 * rows and semantic 'warn's are size hygiene and prune freely.
 *
 * Determinism: keyed on the caller's todayIso (goldens pin todayIso), so
 * never CURRENT_TIMESTAMP / julianday('now') here. The `<` compare is
 * lexicographic, which is correct for both checked_at forms in the table:
 * pipeline writes are date-only 'YYYY-MM-DD', and any legacy
 * DEFAULT-CURRENT_TIMESTAMP 'YYYY-MM-DD HH:MM:SS' rows order the same way
 * ('YYYY-MM-DD HH:MM:SS' sorts after its own 'YYYY-MM-DD' prefix, so a
 * boundary-day timestamp is kept just like a boundary-day date).
 *
 * Accepted downstream behavior shifts (recorded in P7):
 * - theme-review's glossaryRepeatWarnings COUNT becomes a rolling ~30d
 *   window instead of all-time (aligned with its "recent pain" intent);
 * - maintenance's ttlStaleGlossary may fall back to first_seen_date once
 *   a term's audit rows are all pruned. */
export function pruneGlossaryAudit(db: Db, todayIso: string): number {
  return db.prepare(
    `DELETE FROM glossary_audit
      WHERE checked_at < date(?, '-30 days')
        AND NOT (check_type = 'semantic' AND verdict IN ('pass', 'fail'))`,
  ).run(todayIso).changes;
}

export interface ValidateGlossarySummary {
  checked: number;
  /** Audit rows removed by the 30d retention pass (pruneGlossaryAudit). */
  prunedAudit: number;
  retiredByFormOrDedupe: string[];
  warned: string[];
  pendingSemantic: GlossaryRow[];
}

/** The deterministic half (form + dedupe + audit + queue listing). The
 * semantic judge is orchestrator-driven: the caller prompts the LLM
 * over pendingSemantic and commits its verdicts via commitValidation. */
export function runValidateGlossary(db: Db, args: {
  today: string; limit?: number;
}): ValidateGlossarySummary {
  // Retention runs first so today's own writes are never in scope. This
  // rides the daily glossary-validate step, which is replay/dry-run
  // skipped — the prune never executes against golden state.
  const prunedAudit = pruneGlossaryAudit(db, args.today);
  const limit = args.limit ?? 200;
  const rows = db.prepare(
    `SELECT term, aliases_json, quick_def, why_it_matters, canonical_link
       FROM glossary_terms
      WHERE status = 'active'
      ORDER BY term
      LIMIT ?`,
  ).all(limit) as GlossaryRow[];
  const retired: string[] = [];
  const warned: string[] = [];
  for (const row of rows) {
    const verdicts = [formCheck(row), dedupeCheck(db, row)];
    commitValidation(db, { term: row.term, verdicts, today: args.today });
    if (verdicts.some(v => v.verdict === 'fail')) retired.push(row.term);
    else if (verdicts.some(v => v.verdict === 'warn')) warned.push(row.term);
  }
  return {
    checked: rows.length,
    prunedAudit,
    retiredByFormOrDedupe: retired,
    warned,
    pendingSemantic: listPendingSemantic(db, limit),
  };
}
